### ISSUE TITLE
Endpoint Routing Incorrectly Filters Candidates By Media Type

### ISSUE THREAD

**Issue by commonsensesoftware** [dotnet/aspnetcore#33865]
### Describe the bug
_Endpoint Routing_ uses a series of policies to filter candidate actions. When duplicate routes are defined, an appropriate endpoint can still be selected by media type when it shouldn't be. All such cases _should_ result in HTTP 500 with:

`AmbiguousMatchException: The request matched multiple endpoints.`

This was first discovered in microsoft/aspnet-api-versioning#744. Upon further investigation, it was found that the API Versioning **MatcherPolicy** only received a single candidate when it should have received two.

In most API applications, it's unlikely a duplicate route with different media types would be specified, which is probably why it hasn't been previously detected. This is of serious concern to API Versioning because it is very likely that multiple endpoints for different API versions will have the same route template and different media types.

cc: @jacalvar

### To Reproduce

The following will reproduce the issue in a vanilla ASP.NET Core application **without** API Versioning.

```c#
namespace AspNet.Repro
{
    using Microsoft.AspNetCore.Mvc;
    using System;

    public class TestModel
    {
        public int Id { get; set; }
    }

    namespace V1
    {
        [ApiController]
        [Route( "api/[controller]" )]
        public class TestController : ControllerBase
        {
            [HttpPatch( "{id}" )]
            [Consumes( "application/merge-patch+json" )]
            public IActionResult Patch( int id, [FromBody] TestModel model )
            {
                model.Id = id;
                return Ok( model );
            }
        }
    }

    namespace V2
    {
        [ApiController]
        [Route( "api/[controller]" )]
        public class TestController : ControllerBase
        {
            [HttpPatch( "{id}" )]
            // [Consumes( "application/json" )]
            public IActionResult Patch( int id, [FromBody] TestModel model )
            {
                model.Id = id;
                return Ok( model );
            }
        }
    }
}
```

### Behaviors

Sample request:

```http
PATCH api/test/1 HTTP/1.1
Host: localhost:5000
Content-Length: 9

{"id":42}"
```

`Content-Type` varies according to the following table:

| Consumes (on V2) | Content Type | Response | Controller |
| -------------------- | --------------- | ---------- | ----------- |
| `<unspecified>` | `<unspecified>` | 200 | V2.TestController |
| `<unspecified>` | `application/json` | 200 | V2.TestController |
| `<unspecified>` | `application/merge-patch+json` | 200 | V1.TestController |
| `application/json` | `<unspecified>` | 200 | V2.TestController |
| `application/json` | `application/json` | 200 | V2.TestController |
| `application/json` | `application/merge-patch+json` | 500 | `<exception>` |

### Further technical details
- ASP.NET Core version
- Include the output of `dotnet --info`
- The IDE (VS / VS Code/ VS4Mac) you're running on, and its version

```
.NET SDK (reflecting any global.json):
 Version:   5.0.301
 Commit:    ef17233f86

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19042
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\5.0.301\

Host (useful for support):
  Version: 5.0.7
  Commit:  556582d964

.NET SDKs installed:
  5.0.301 [C:\Program Files\dotnet\sdk]
```

**Comment by javiercn** [comment]
@commonsensesoftware thanks for bringing this up Chris.

If I understand correctly, you are suggesting that the consumers matcher policy is incorrectly filtering out one of the endpoints which results in the route being matched instead of a 500 (ambiguous route) being produced?

I think I need to think a bit more about this, since I'm not super familiar with the behavior here. We definitely support the media type and the subtype, but I don't remember how we deal with suffixes. There are three possibilities here:
* We treat everything after the "/" as a single unit (meaning application/json is different from application/merge-patch+json).
* We treat application/merge-patch+json as compatible with application/json.
* We treat application/json as compatible with application/merge-patch+json.
* We treat application/merge-patch+json as more specific than application/json

I'll have to read the details on how this worked and remember what was going on here. This is potentially a breaking change, so we'll have to evaluate our options here.

It'll help if you can point out which endpoint matches when the response is 200, since that will help wrap my head around this.

**Comment by commonsensesoftware** [comment]
Thanks for the quick reply Javier. Per your request, I've updated the table with the selected controller.

For clarity, I _think_ the selection criteria you've outlined is correct and matches the behavior. Matching unspecified media types seems not entirely obvious, but I can understand that decision. The primary issue is the only single candidate is being discovered. Both candidates _should_ be discovered in all scenarios, regardless of the selected matches. Candidates should be marked as invalid based on media type. This should be the same behavior of HTTP methods or other criteria.

One of the limitations addressed in _Endpoint Routing_ was to separate candidate discovery from matching/selection. This gives other policies an opportunity to make a decision.

**Comment by javiercn** [comment]
@commonsensesoftware I agree in general.

I suspect that the ConsumesMatcherPolicy is involved here through the INodeBuilderPolicy, and I'm wondering if the behavior is the same or is different when you have a dynamic endpoint in the list of candidates (Having a dynamic endpoint prevents the node policy builder from taking effect).

If you want to dig into this, you can try it out by creating an additional endpoint that matches the route and that has an attribute that implements `IDynamicEndpointMetadata` and returns true. You can make that endpoint only consume `xml` or something like that, and that should be enough for it to get filtered out but prevent the INodeBuilderPolicy from applying.

**Comment by commonsensesoftware** [comment]
Thanks @javiercn. I spent quite a bit of time spelunking back through the _Endpoint Routing_ infrastructure. In short, this is **not** a bug and is the expected behavior from a vanilla routing perspective. This results weren't what I would have been expecting, but upon further analysis, it does seem to be correct.

The fundamental issue is that candidates are _eliminated_ not merely _invalidated_ by earlier **MatcherPolicy** instances such as **HttpMethodMatcherPolicy** and **ConsumesMatcherPolicy**. This works fine and well for out-of-the-box scenarios, but is definitely a problem for API Versioning and is unexpected in its current design. This appears to be my misunderstanding. The basic design has always been to merely disambiguate otherwise ambiguous routes by way of an API version. Things fall down in the case of 405 and 415 responses. I _think_ I largely had 405 solved, which is why this hasn't come up before. In the case of 415, things simply don't work right.

After printing out an existing DFA, I created this simple variant, which I _believe_ illustrates what needs to happen.

<details>
<summary>(Show source)

<img src="http://www.plantuml.com/plantuml/png/ZP4zRy8m48Pt_mgBaGexTi97AhK2jQeO6RAr7Lomm9LZhCHCYFzUNuWSDIAA9tJtl7vEl5dNJbJkKcfqKCTAs1D-_zZWAqhmfnO_KhzCwEJRJxDNp13hbi8g00m0Hxm3JjQENWaXD-0SU7h_oL0QOemadcPXvisSGcmUbjkYs3tZtQPusteu_tVyVWT_cLu838kG5DPgjHTEdGqjPNMKColS_lJqMv-DJoz6bNuQSJGNbq4qfLE_NCQ88Tg-Nw5LpB_B89iFz9F76FBcGoO32W2n3lQmvBOKeD7DXeV1pNjrCXxVvmsjfJbSIlq7" />
</summary>

```dot
digraph DFA {
0 [label="/"]
0 -> 1
1 [label="/api/"]
1 -> 2
2 [label="/test/{...}/"]
2 -> 3
2 -> 4
3 [label="1.0"]
4 [label="2.0"]
3 -> 5 [label="HTTP: PATCH"]
5 [label="/api/test/ \nHTTP: PATCH"]
5 -> 6 [label="application/merge-patch+json"]
6 [label="/api/test/ \nHTTP: PATCH \napplication/merge-patch+json"]
5 -> 7 [label="*/*"]
7 [label="/api/test/ \nHTTP: PATCH \n*/*"]
5 -> 8
8 [label="/api/test/ \nHTTP: PATCH"]
4 -> 9 [label="HTTP: PATCH"]
9 [label="/api/test/ \nHTTP: PATCH"]
9 -> 10 [label="HTTP: PATCH"]
10 [label="/api/test/ \nHTTP: PATCH"]
9 -> 11 [label="application/json"]
11 [label="/api/test/ \nHTTP: PATCH \napplication/json"]
9 -> 12 [label="*/*"]
12 [label="/api/test/ \nHTTP: PATCH \n*/*"]
}
```

</details>

This is **not** a trivial change and would require implementing **INodeBuilderPolicy**. Back when I was working with the team to vet the design, we agreed that **IEndpointSelectorPolicy** would have been sufficient. In most cases it is, but not for 405 or 415 because they eliminate candidates before API Versioning sees them, which results in the incorrect response.

Assuming that I can make **INodeBuilderPolicy** work this way, I _think_ it's the way to go. Eliminating candidates by API version  ahead of **HttpMethodMatcherPolicy** and **ConsumesMatcherPolicy** _should_ yield the correct behavior. I also think there is a chance that versioned endpoints will be resolved faster (we'll see). This will unfortunately mean that the order that the **ApiVersionMatcherPolicy** runs in will be important.

It's probably going to take a bit to flush out whether this is the right approach and will work. If it does, then I'll close this issue out. If it doesn't, then I'll elaborate further so we can look at queueing up whatever future feature may be necessary to enable the right behavior. If you have any other suggestions or ideas, I'm open to them. 😃

**Comment by commonsensesoftware** [comment]
After a **lot** of investigation and diving _into the weeds_ of the routing system, I can confirm that my last assessment was correct. _Endpoint Routing_ is working as expected. **This is not a bug**. I did, however, discover that in order to make API Versioning work properly with these behaviors, it will need to implement `INodeBuilderPolicy` and return a specialized `PolicyJumpTable` which runs in front of the default policies. Understanding how node policies work isn't terribly difficult. Implementing said policy is not for the faint of heart. It's arguably TMI for this issue, but it will now effectively work as above. The next major release for API Versioning will have the necessary changes in place.

Thanks for the assist and taking the time to consider looking into this more deeply. I feel we can close this out since there's nothing more to do on your side and I've ironed out what must be done in API Versioning.

cc: @javiercn
